### PR TITLE
gh-96670: Raise SyntaxError when parsing NULL bytes

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-96670

### CODE REVIEW

File: Parser/tokenizer.c [hunk 1]
```diff
@@ -829,12 +844,25 @@ tok_readline_raw(struct tok_state *tok)
...
-                                              tok->fp, NULL);
+        memset(tok->inp, 0, BUFSIZ);
+        int n_chars = (int)(tok->end - tok->inp);
+        char *line = Py_UniversalNewlineFgets(tok->inp, n_chars, tok->fp, NULL);
         if (line == NULL) {
             return 1;
         }
+        
+        // Find the first non null character starting from the right of tok->inp
+        char *last_char = tok->inp + n_chars - 1;
```
Comment (pablogsal) [inline Parser/tokenizer.c:855]: is a bit unfortunate that we need to do this to know the last character that was ingested, but the alternative is reimplementing `Py_UniversalNewlineFgets` so it tells us how many chars actually added to the buffer

Comment (pablogsal) [reply Parser/tokenizer.c]: I have added a commit adding a new interface for `Py_UniversalNewlineFgets` that gives us the size for consideration

**Review by gvanrossum — COMMENTED** [review]
What's up with python3.12.abi.new having 22,644 additions?

File: Parser/tokenizer.c [hunk 2]
```diff
@@ -1075,6 +1080,12 @@ tok_nextc(struct tok_state *tok)
             return EOF;
         }
         tok->line_start = tok->cur;
+
+        if (contains_null_bytes(tok->line_start, tok->inp - tok->line_start)) {
+            syntaxerror(tok, "source code string cannot contain null bytes");
```
Comment (gvanrossum) [inline Parser/tokenizer.c]: We get here when reading from a file too, right?
```suggestion
            syntaxerror(tok, "source code cannot contain null bytes");
```

**Review by gvanrossum — COMMENTED** [review]
LG except for one word in an error message, I think.

**Comment by pablogsal** [comment]
> What's up with python3.12.abi.new having 22,644 additions?

Wow, that's some error. Let me clean that up

File: Parser/tokenizer.c (hunk 2 above)
Comment (pablogsal) [reply Parser/tokenizer.c]: Done!

File: Objects/fileobject.c [hunk 3]
```diff
@@ -239,7 +239,7 @@ _PyLong_FileDescriptor_Converter(PyObject *o, void *ptr)
 ** identically.
 */
 char *
-Py_UniversalNewlineFgets(char *buf, int n, FILE *stream, PyObject *fobj)
+_Py_UniversalNewlineFgetsWithSize(char *buf, int n, FILE *stream, PyObject *fobj, size_t* size)
```
Comment (lysnikolaou) [inline Objects/fileobject.c:234]: Nit: Maybe move the docstring down to the new definition of `Py_UniversalNewlineFgets`?

Comment (pablogsal) [reply Objects/fileobject.c:234]: Makes sense, will do

**Comment by gvanrossum** [comment]
Ugh, force-push. :-(

**Review by gvanrossum — COMMENTED** [review]
I see nothing else wrong, let's wait for Lysandros' LGTM.

**Review by lysnikolaou — APPROVED** [review]
LGTM as well!

**Comment by mdboom** [comment]
I was going to ask "doesn't this break things for encodings where NULLs are normal, like UTF-16", but then realized we never supported that anyway: 

PEP 253: "It does not include encodings which use two or more bytes for all characters like e.g. UTF-16. The reason for this is to keep the encoding detection algorithm in the tokenizer simple."

In that sense, we get a better error message for UTF-16 files with this PR than before.  Prior, it would silently fail (behave as if an empty file), and with this PR:

```
  File "/home/mdboom/tmp/utf16-manual.py", line 1
    #
SyntaxError: source code cannot contain null bytes  File "x.py", line 1
```

**Comment by pablogsal** [comment]
Landing, thank you everyone for your great review and comments! And also thank @mdboom for checking the UTF-16, that was a good consideration :)

P.S. Sorry for the force push :S

**Pull request merged**